experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyField.ts

```diff
@@ -254,26 +256,122 @@ export class LazySequence<TTypes extends AllowedTypes>
+		this.insertAt(0, value);
+	}
+
+	public insertAtEnd(value: FlexibleNodeContent<TTypes>[]): void {
+		this.insertAt(this.length, value);
+	}
+
+	public removeAt(index: number): void {
+		const fieldEditor = this.sequenceEditor();
+		fieldEditor.delete(index, 1);
+	}
+
+	public removeRange(start?: number, end?: number): void {
+		const fieldEditor = this.sequenceEditor();
+		const { length } = this;
+		const removeStart = start ?? 0;
+		const removeEnd = Math.min(length, end ?? length);
+		assertValidRangeIndices(removeStart, removeEnd, this);
+		fieldEditor.delete(removeStart, removeEnd - removeStart);
+	}
+
+	public moveToStart(sourceStart: number, sourceEnd: number): void;
+	public moveToStart<TTypesSource extends AllowedTypes>(
+		sourceStart: number,
+		sourceEnd: number,
+		source: Sequence<CheckTypesOverlap<TTypesSource, TTypes>>,
+	): void;
+	public moveToStart<TTypesSource extends AllowedTypes>(
+		sourceStart: number,
+		sourceEnd: number,
+		source?: Sequence<CheckTypesOverlap<TTypesSource, TTypes>>,
+	): void {
+		this.moveToIndexI(0, sourceStart, sourceEnd, source);
+	}
+
+	public moveToEnd(sourceStart: number, sourceEnd: number): void;
+	public moveToEnd<TTypesSource extends AllowedTypes>(
+		sourceStart: number,
+		sourceEnd: number,
+		source: Sequence<CheckTypesOverlap<TTypesSource, TTypes>>,
+	): void;
+	public moveToEnd<TTypesSource extends AllowedTypes>(
+		sourceStart: number,
+		sourceEnd: number,
+		source?: Sequence<CheckTypesOverlap<TTypesSource, TTypes>>,
+	): void {
+		this.moveToIndexI(this.length, sourceStart, sourceEnd, source);
+	}
+
+	public moveToIndex(index: number, sourceStart: number, sourceEnd: number): void;
+	public moveToIndex<TTypesSource extends AllowedTypes>(
+		index: number,
+		sourceStart: number,
+		sourceEnd: number,
+		source: Sequence<CheckTypesOverlap<TTypesSource, TTypes>>,
+	): void;
+	public moveToIndex<TTypesSource extends AllowedTypes>(
+		index: number,
+		sourceStart: number,
+		sourceEnd: number,
+		source?: Sequence<CheckTypesOverlap<TTypesSource, TTypes>>,
+	): void {
+		this.moveToIndexI(index, sourceStart, sourceEnd, source);
+	}
+
+	private moveToIndexI<TTypesSource extends AllowedTypes>(
+		index: number,
+		sourceStart: number,
+		sourceEnd: number,
+		source?: Sequence<CheckTypesOverlap<TTypesSource, TTypes>>,
+	): void {
+		const sourceField = source !== undefined ? (this.isSameAs(source) ? this : source) : this;
+		assertValidRangeIndices(sourceStart, sourceEnd, sourceField);
+		if (this.schema.types !== undefined && sourceField !== this) {
+			for (let i = sourceStart; i < sourceEnd; i++) {
+				const sourceNode = sourceField.at(sourceStart);
+				if (!this.schema.types.has(sourceNode.schema.name)) {
+					throw new Error("Type in source sequence is not allowed in destination.");
+				}
+			}
+		}
+		const count = sourceEnd - sourceStart;
+		let destinationIndex = index;
+		if (sourceField === this) {
+			destinationIndex -= count;
```

Comment:
Shouldn't we only do this if the index is greater than source start? What happens if the index is in the middle of the range - should we only subtract part of count?

Reply:
no (discussed offline). I'll add a comment, though.